server.go

```diff
@@ -114,10 +116,10 @@ func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (
 			server.Unlock()
 			return nil, false, errPoolLimit
 		}
-		if n > 0 {
-			socket = server.unusedSockets[n-1]
-			server.unusedSockets[n-1] = nil // Help GC.
-			server.unusedSockets = server.unusedSockets[:n-1]
+		if n > 0 && len(server.liveSockets) >= minPoolLimit {
+			socket = server.unusedSockets[0]
+			server.unusedSockets[0] = nil // Help GC.
+			server.unusedSockets = server.unusedSockets[1:]
```

Comment:
I'm not exactly sure how this is implemented, but presumably this is going to cause a full shift of all elements. We might consider switching this from a slice to a linked list. Not a huge deal but I thought I would mention it.